sphinx/util/inspect.py

```diff
@@ -598,15 +599,19 @@ def signature(subject: Callable, bound_method: bool = False, type_aliases: dict
                              __validate_parameters__=False)
 
 
-def evaluate_signature(sig: inspect.Signature, globalns: dict | None = None,
-                       localns: dict | None = None,
+def evaluate_signature(sig: inspect.Signature, globalns: dict[str, object] | None = None,
+                       localns: dict[str, object] | None = None,
                        ) -> inspect.Signature:
     """Evaluate unresolved type annotations in a signature object."""
-    def evaluate_forwardref(ref: ForwardRef, globalns: dict, localns: dict) -> Any:
+    def evaluate_forwardref(
+        ref: ForwardRef, globalns: dict[str, object] | None, localns: dict[str, object] | None,
```

Comment:
This cluster of three methods (`evaluate_signature`, `evaluate_forwardref`, `evaluate`) is the only part I'm uncertain about so far, because `ForwardRef._evaluate` seems to be an undocumented method.

Reply:
yeah it's a weird one.

here's the signature of `ForwardRef._evaluate`

```python
    def _evaluate(self, globalns, localns, recursive_guard):
        if self.__forward_arg__ in recursive_guard:
            return self
        if not self.__forward_evaluated__ or localns is not globalns:
            if globalns is None and localns is None:
                globalns = localns = {}
            elif globalns is None:
                globalns = localns
            elif localns is None:
                localns = globalns
            if self.__forward_module__ is not None:
                globalns = getattr(
                    sys.modules.get(self.__forward_module__, None), '__dict__', globalns
                )
            type_ = _type_check(
                eval(self.__forward_code__, globalns, localns),
                "Forward references must evaluate to types.",
                is_argument=self.__forward_is_argument__,
                allow_special_forms=self.__forward_is_class__,
            )
            self.__forward_value__ = _eval_type(
                type_, globalns, localns, recursive_guard | {self.__forward_arg__}
            )
            self.__forward_evaluated__ = True
        return self.__forward_value__
```

and here's the type stub of the `eval` method where those args get used:

```python
def eval(
    __source: str | ReadableBuffer | CodeType,
    __globals: dict[str, Any] | None = None,
    __locals: Mapping[str, object] | None = None,
) -> Any: ...
```

so my type signature (`globalns: dict[str, object] | None, localns: dict[str, object] | None`) is consistent with the usage in `eval`

Reply:
I'd like to suggest changing `object` to `Any` since the latest typeshed uses `dict[str, Any]`: https://github.com/python/typeshed/blob/49b1a1a96a90946ff8885792eec30acb5cf39af0/stdlib/typing.pyi#L929

Reply:
> I'd like to suggest changing `object` to `Any` since the latest typeshed uses `dict[str, Any]`: https://github.com/python/typeshed/blob/49b1a1a96a90946ff8885792eec30acb5cf39af0/stdlib/typing.pyi#L929

done!